bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/handler/AbstractMieleThingHandler.java

```diff
@@ -0,0 +1,292 @@
+import org.openhab.binding.mielecloud.internal.webservice.api.PowerStatus;
+import org.openhab.binding.mielecloud.internal.webservice.api.ProgramStatus;
+import org.openhab.binding.mielecloud.internal.webservice.api.TransitionState;
+import org.openhab.binding.mielecloud.internal.webservice.api.json.ProcessAction;
+import org.openhab.binding.mielecloud.internal.webservice.api.json.StateType;
+import org.openhab.binding.mielecloud.internal.webservice.exception.TooManyRequestsException;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.BridgeHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract base class for all Miele thing handlers.
+ *
+ * @author Roland Edelhoff - Initial contribution
+ * @author Björn Lange - Add channel state wrappers
+ */
+@NonNullByDefault
+public abstract class AbstractMieleThingHandler extends BaseThingHandler {
+    protected final ActionStateFetcher actionFetcher;
+    protected DeviceState latestDeviceState = new DeviceState(getDeviceId(), null);
+    protected TransitionState latestTransitionState = new TransitionState(null, latestDeviceState);
+    protected ActionsState latestActionsState = new ActionsState(getDeviceId(), null);
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    /**
+     * Creates a new {@link AbstractMieleThingHandler}.
+     *
+     * @param thing The thing to handle.
+     */
+    public AbstractMieleThingHandler(Thing thing) {
+        super(thing);
+        this.actionFetcher = new ActionStateFetcher(this::getWebservice, scheduler);
+    }
+
+    private Optional<MieleBridgeHandler> getMieleBridgeHandler() {
+        Bridge bridge = getBridge();
+        if (bridge == null) {
+            return Optional.empty();
+        }
+
+        BridgeHandler handler = bridge.getHandler();
+        if (handler == null || !(handler instanceof MieleBridgeHandler)) {
+            return Optional.empty();
+        }
+
+        return Optional.of((MieleBridgeHandler) handler);
+    }
+
+    protected MieleWebservice getWebservice() {
+        return getMieleBridgeHandler().map(MieleBridgeHandler::getWebservice)
+                .orElse(UnavailableMieleWebservice.INSTANCE);
+    }
+
+    @Override
+    public void initialize() {
+        getWebservice().dispatchDeviceState(getDeviceId());
+
+        // If no device state update was received so far, set the device to OFFLINE.
+        if (getThing().getStatus() == ThingStatus.INITIALIZING) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```

Comment:
In case of a COMMUNICATION_ERROR, a message text to tell the user what happened is highly recommended.
In the situation here, I feel that you probably only want to say OFFLINE without actually knowing the exact reason for it, don't you?

Reply:
I must admit, if I think closely of this we take a slight guess here. When the thing is initialized, it calls out to the bridge to synchronously dispatch the current state of all things. This update is processed by the `AbstractMieleThingHandler` and the status is set accordingly. If the status is not set then either there is no connection to the Miele cloud or the thing was removed while the binding was inactive. The guess is that it is far more uncommon that a device was removed in the cloud than a communication problem is. You are right setting it to just `OFFLINE` seems to be the better option here.